File: ios/PttWire/Sources/PttWire/SFrame.swift
import CryptoKit
import Foundation

public let sframeAes128GcmSha256128: UInt16 = 0x0004

public protocol SFrameCounterStore: AnyObject {
    /// Persists and returns an unused counter before any ciphertext is emitted.
    func takeNext(kid: UInt64) throws -> UInt64
}

public final class MemorySFrameCounterStore: SFrameCounterStore {
    private var next: [UInt64: UInt64] = [:]
    private let lock = NSLock()

    public init() {}

    public func takeNext(kid: UInt64) throws -> UInt64 {
        lock.lock()
        defer { lock.unlock() }
        let value = next[kid, default: 0]
        guard value != UInt64.max else { throw SFrameError.counterExhausted }
        next[kid] = value + 1
        return value
    }
}

public final class SFrameEncryptor {
    private let kid: UInt64
    private let key: SymmetricKey
    private let salt: Data
    private let counters: SFrameCounterStore

    public init(kid: UInt64, baseKey: Data, counters: SFrameCounterStore) throws {
        guard !baseKey.isEmpty else { throw SFrameError.invalidKey }
        self.kid = kid
        self.key = SymmetricKey(data: derive(baseKey, info: label("SFrame 1.0 Secret key ", kid), count: 16))
        self.salt = derive(baseKey, info: label("SFrame 1.0 Secret salt ", kid), count: 12)
        self.counters = counters
    }

    public func encrypt(metadata: Data, plaintext: Data) throws -> Data {
        try encrypt(counter: counters.takeNext(kid: kid), metadata: metadata, plaintext: plaintext)
    }

    func encrypt(counter: UInt64, metadata: Data, plaintext: Data) throws -> Data {
        let header = SFrameHeader.encode(kid: kid, counter: counter)
        let sealed = try AES.GCM.seal(
            plaintext,
            using: key,
            nonce: AES.GCM.Nonce(data: nonce(salt: salt, counter: counter)),
            authenticating: header + metadata
        )
        return header + sealed.ciphertext + sealed.tag
    }
}

public final class SFrameDecryptor {
    private struct Material {
        let key: SymmetricKey
        let salt: Data
    }

    private var keys: [UInt64: Material] = [:]
    private var replay: [UInt64: ReplayWindow] = [:]
    private let lock = NSLock()

    public init() {}

    public func addKey(kid: UInt64, baseKey: Data) throws {
        guard !baseKey.isEmpty else { throw SFrameError.invalidKey }
        lock.lock()
        defer { lock.unlock() }
        keys[kid] = Material(
            key: SymmetricKey(data: derive(baseKey, info: label("SFrame 1.0 Secret key ", kid), count: 16)),
            salt: derive(baseKey, info: label("SFrame 1.0 Secret salt ", kid), count: 12)
        )
        replay.removeValue(forKey: kid)
    }

    public func removeKey(kid: UInt64) {
        lock.lock()
        defer { lock.unlock() }
        keys.removeValue(forKey: kid)
        replay.removeValue(forKey: kid)
    }

    public func decrypt(metadata: Data, frame: Data) throws -> Data {
        lock.lock()
        defer { lock.unlock() }
        let parsed = try SFrameHeader.parse(frame)
        guard let material = keys[parsed.kid] else { throw SFrameError.unknownKey }
        var window = replay[parsed.kid, default: ReplayWindow()]
        guard window.acceptable(parsed.counter) else { throw SFrameError.replay }
        guard frame.count >= parsed.length + 16 else { throw SFrameError.malformedHeader }
        let ciphertext = frame[parsed.length..<(frame.count - 16)]
        let tag = frame.suffix(16)
        do {
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: nonce(salt: material.salt, counter: parsed.counter)),
                ciphertext: ciphertext,
                tag: tag
            )
            let plaintext = try AES.GCM.open(
                box,
                using: material.key,
                authenticating: frame.prefix(parsed.length) + metadata
            )
            window.mark(parsed.counter)
            replay[parsed.kid] = window
            return plaintext
        } catch {
            throw SFrameError.authenticationFailed
        }
    }
}

public enum SFrameError: Error, Equatable {
    case invalidKey
    case counterExhausted
    case malformedHeader
    case unknownKey
    case authenticationFailed
    case replay
}

private struct ParsedSFrameHeader {
    let kid: UInt64
    let counter: UInt64
    let length: Int
}

private enum SFrameHeader {
    static func encode(kid: UInt64, counter: UInt64) -> Data {
        let kidBytes = compact(kid)
        let counterBytes = compact(counter)
        let kidExtended = kid >= 8
        let counterExtended = counter >= 8
        var config: UInt8 = 0
        if kidExtended {
            config |= 0x80 | UInt8((kidBytes.count - 1) << 4)
        } else {
            config |= UInt8(kid << 4)
        }
        if counterExtended {
            config |= 0x08 | UInt8(counterBytes.count - 1)
        } else {
            config |= UInt8(counter)
        }
        var output = Data([config])
        if kidExtended { output.append(contentsOf: kidBytes) }
        if counterExtended { output.append(contentsOf: counterBytes) }
        return output
    }

    static func parse(_ frame: Data) throws -> ParsedSFrameHeader {
        guard let config = frame.first else { throw SFrameError.malformedHeader }
        let kidExtended = config & 0x80 != 0
        let counterExtended = config & 0x08 != 0
        let kidLength = kidExtended ? Int((config >> 4) & 0x07) + 1 : 0
        let counterLength = counterExtended ? Int(config & 0x07) + 1 : 0
        let length = 1 + kidLength + counterLength
        guard frame.count >= length else { throw SFrameError.malformedHeader }
        var offset = 1
        let kid: UInt64
        if kidExtended {
            kid = try decode(frame, offset: offset, length: kidLength)
            offset += kidLength
        } else {
            kid = UInt64((config >> 4) & 0x07)
        }
        let counter = counterExtended
            ? try decode(frame, offset: offset, length: counterLength)
            : UInt64(config & 0x07)
        guard (!kidExtended || kid >= 8), (!counterExtended || counter >= 8) else {
            throw SFrameError.malformedHeader
        }
        return ParsedSFrameHeader(kid: kid, counter: counter, length: length)
    }

    private static func compact(_ value: UInt64) -> [UInt8] {
        let bytes = withUnsafeBytes(of: value.bigEndian) { Array($0) }
        let first = bytes.firstIndex(where: { $0 != 0 }) ?? 7
        return Array(bytes[first...])
    }

    private static func decode(_ data: Data, offset: Int, length: Int) throws -> UInt64 {
        guard (1...8).contains(length), !(length > 1 && data[offset] == 0) else {
            throw SFrameError.malformedHeader
        }
        return data[offset..<(offset + length)].reduce(0) { ($0 << 8) | UInt64($1) }
    }
}

private struct ReplayWindow {
    private var initialized = false
    private var highest: UInt64 = 0
    private var low: UInt64 = 0
    private var high: UInt64 = 0

    func acceptable(_ counter: UInt64) -> Bool {
        guard initialized, counter <= highest else { return true }
        let distance = highest - counter
        guard distance < 128 else { return false }
        if distance < 64 { return low & (1 << distance) == 0 }
        return high & (1 << (distance - 64)) == 0
    }

    mutating func mark(_ counter: UInt64) {
        guard initialized else {
            initialized = true
            highest = counter
            low = 1
            return
        }
        if counter > highest {
            let distance = counter - highest
            if distance >= 128 {
                low = 1
                high = 0
            } else if distance >= 64 {
                high = low << (distance - 64)
                low = 1
            } else {
                high = (high << distance) | (low >> (64 - distance))
                low = (low << distance) | 1
            }
            highest = counter
        } else {
            let distance = highest - counter
            if distance < 64 { low |= 1 << distance }
            else { high |= 1 << (distance - 64) }
        }
    }
}

private func label(_ prefix: String, _ kid: UInt64) -> Data {
    var output = Data(prefix.utf8)
    var bigEndianKid = kid.bigEndian
    output.append(Data(bytes: &bigEndianKid, count: 8))
    var suite = sframeAes128GcmSha256128.bigEndian
    output.append(Data(bytes: &suite, count: 2))
    return output
}

private func derive(_ baseKey: Data, info: Data, count: Int) -> Data {
    let key = HKDF<SHA256>.deriveKey(
        inputKeyMaterial: SymmetricKey(data: baseKey),
        salt: Data(),
        info: info,
        outputByteCount: count
    )
    return key.withUnsafeBytes { Data($0) }
}

private func nonce(salt: Data, counter: UInt64) -> Data {
    var output = Array(salt)
    let counterBytes = withUnsafeBytes(of: counter.bigEndian) { Array($0) }
    for index in counterBytes.indices { output[output.count - 8 + index] ^= counterBytes[index] }
    return Data(output)
}

File: android/media/src/main/kotlin/app/ptt/media/SFrame.kt
package app.ptt.media

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

const val SFRAME_AES_128_GCM_SHA256_128: Int = 0x0004

interface SFrameCounterStore {
    /** Persists and returns an unused counter before any ciphertext is emitted. */
    fun takeNext(kid: ULong): ULong
}

class MemorySFrameCounterStore : SFrameCounterStore {
    private val next = mutableMapOf<ULong, ULong>()

    @Synchronized
    override fun takeNext(kid: ULong): ULong {
        val value = next[kid] ?: 0uL
        check(value != ULong.MAX_VALUE) { "SFrame counter exhausted" }
        next[kid] = value + 1u
        return value
    }
}

class SFrameEncryptor(
    private val kid: ULong,
    baseKey: ByteArray,
    private val counters: SFrameCounterStore,
) {
    private val key: ByteArray
    private val salt: ByteArray

    init {
        require(baseKey.isNotEmpty()) { "invalid SFrame base key" }
        key = derive(baseKey, label("SFrame 1.0 Secret key ", kid), 16)
        salt = derive(baseKey, label("SFrame 1.0 Secret salt ", kid), 12)
    }

    fun encrypt(metadata: ByteArray, plaintext: ByteArray): ByteArray =
        encryptWithCounter(counters.takeNext(kid), metadata, plaintext)

    internal fun encryptWithCounter(
        counter: ULong,
        metadata: ByteArray,
        plaintext: ByteArray,
    ): ByteArray {
        val header = SFrameHeader.encode(kid, counter)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(
            Cipher.ENCRYPT_MODE,
            SecretKeySpec(key, "AES"),
            GCMParameterSpec(128, nonce(salt, counter)),
        )
        cipher.updateAAD(header + metadata)
        return header + cipher.doFinal(plaintext)
    }
}

class SFrameDecryptor {
    private data class KeyMaterial(val key: ByteArray, val salt: ByteArray)

    private val keys = mutableMapOf<ULong, KeyMaterial>()
    private val replay = mutableMapOf<ULong, ReplayWindow>()

    @Synchronized
    fun addKey(kid: ULong, baseKey: ByteArray) {
        require(baseKey.isNotEmpty()) { "invalid SFrame base key" }
        keys[kid] =
            KeyMaterial(
                derive(baseKey, label("SFrame 1.0 Secret key ", kid), 16),
                derive(baseKey, label("SFrame 1.0 Secret salt ", kid), 12),
            )
        replay.remove(kid)
    }

    @Synchronized
    fun removeKey(kid: ULong) {
        keys.remove(kid)
        replay.remove(kid)
    }

    @Synchronized
    fun decrypt(metadata: ByteArray, frame: ByteArray): ByteArray {
        val parsed = SFrameHeader.parse(frame)
        val material = keys[parsed.kid] ?: throw SFrameException.UnknownKey
        val window = replay.getOrPut(parsed.kid) { ReplayWindow() }
        if (!window.acceptable(parsed.counter)) throw SFrameException.Replay
        if (frame.size < parsed.length + 16) throw SFrameException.MalformedHeader
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(
            Cipher.DECRYPT_MODE,
            SecretKeySpec(material.key, "AES"),
            GCMParameterSpec(128, nonce(material.salt, parsed.counter)),
        )
        cipher.updateAAD(frame.copyOfRange(0, parsed.length) + metadata)
        val plaintext =
            try {
                cipher.doFinal(frame, parsed.length, frame.size - parsed.length)
            } catch (error: Exception) {
                throw SFrameException.AuthenticationFailed(error)
            }
        window.mark(parsed.counter)
        return plaintext
    }
}

sealed class SFrameException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    data object MalformedHeader : SFrameException("malformed or non-canonical SFrame header")
    data object UnknownKey : SFrameException("SFrame key is not available")
    data object Replay : SFrameException("SFrame counter was replayed or is outside the replay window")
    class AuthenticationFailed(cause: Throwable) : SFrameException("SFrame authentication failed", cause)
}

private data class ParsedSFrameHeader(val kid: ULong, val counter: ULong, val length: Int)

private object SFrameHeader {
    fun encode(kid: ULong, counter: ULong): ByteArray {
        val kidBytes = compact(kid)
        val counterBytes = compact(counter)
        val kidExtended = kid >= 8u
        val counterExtended = counter >= 8u
        var config = 0
        config =
            if (kidExtended) config or 0x80 or ((kidBytes.size - 1) shl 4)
            else config or (kid.toInt() shl 4)
        config =
            if (counterExtended) config or 0x08 or (counterBytes.size - 1)
            else config or counter.toInt()
        val output = ByteArrayOutputStream(17)
        output.write(config)
        if (kidExtended) output.write(kidBytes)
        if (counterExtended) output.write(counterBytes)
        return output.toByteArray()
    }

    fun parse(frame: ByteArray): ParsedSFrameHeader {
        if (frame.isEmpty()) throw SFrameException.MalformedHeader
        val config = frame[0].toInt() and 0xff
        val kidExtended = config and 0x80 != 0
        val counterExtended = config and 0x08 != 0
        val kidLength = if (kidExtended) ((config ushr 4) and 0x07) + 1 else 0
        val counterLength = if (counterExtended) (config and 0x07) + 1 else 0
        val length = 1 + kidLength + counterLength
        if (frame.size < length) throw SFrameException.MalformedHeader
        var offset = 1
        val kid =
            if (kidExtended) decode(frame, offset, kidLength).also { offset += kidLength }
            else ((config ushr 4) and 0x07).toULong()
        val counter =
            if (counterExtended) decode(frame, offset, counterLength)
            else (config and 0x07).toULong()
        if ((kidExtended && kid < 8u) || (counterExtended && counter < 8u)) {
            throw SFrameException.MalformedHeader
        }
        return ParsedSFrameHeader(kid, counter, length)
    }

    private fun compact(value: ULong): ByteArray {
        val bytes = ByteBuffer.allocate(8).putLong(value.toLong()).array()
        val first = bytes.indexOfFirst { it.toInt() != 0 }.let { if (it == -1) 7 else it }
        return bytes.copyOfRange(first, 8)
    }

    private fun decode(bytes: ByteArray, offset: Int, length: Int): ULong {
        if (length !in 1..8 || (length > 1 && bytes[offset].toInt() == 0)) {
            throw SFrameException.MalformedHeader
        }
        var result = 0uL
        repeat(length) { result = (result shl 8) or (bytes[offset + it].toUByte().toULong()) }
        return result
    }
}

private class ReplayWindow {
    private var initialized = false
    private var highest = 0uL
    private var low = 0uL
    private var high = 0uL

    fun acceptable(counter: ULong): Boolean {
        if (!initialized || counter > highest) return true
        val distance = highest - counter
        if (distance >= 128u) return false
        val bit = distance.toInt()
        return if (bit < 64) low and (1uL shl bit) == 0uL
        else high and (1uL shl (bit - 64)) == 0uL
    }

    fun mark(counter: ULong) {
        if (!initialized) {
            initialized = true
            highest = counter
            low = 1u
            return
        }
        if (counter > highest) {
            val distance = counter - highest
            when {
                distance >= 128u -> {
                    low = 1u
                    high = 0u
                }
                distance >= 64u -> {
                    high = low shl (distance.toInt() - 64)
                    low = 1u
                }
                else -> {
                    val shift = distance.toInt()
                    high = (high shl shift) or (low shr (64 - shift))
                    low = (low shl shift) or 1u
                }
            }
            highest = counter
        } else {
            val distance = (highest - counter).toInt()
            if (distance < 64) low = low or (1uL shl distance)
            else high = high or (1uL shl (distance - 64))
        }
    }
}

private fun label(prefix: String, kid: ULong): ByteArray =
    prefix.encodeToByteArray() +
        ByteBuffer.allocate(8).putLong(kid.toLong()).array() +
        byteArrayOf(0, SFRAME_AES_128_GCM_SHA256_128.toByte())

private fun derive(baseKey: ByteArray, info: ByteArray, length: Int): ByteArray {
    val extract = Mac.getInstance("HmacSHA256")
    extract.init(SecretKeySpec(ByteArray(32), "HmacSHA256"))
    val pseudoRandomKey = extract.doFinal(baseKey)
    val expand = Mac.getInstance("HmacSHA256")
    expand.init(SecretKeySpec(pseudoRandomKey, "HmacSHA256"))
    expand.update(info)
    expand.update(1)
    return expand.doFinal().copyOf(length)
}

private fun nonce(salt: ByteArray, counter: ULong): ByteArray {
    val result = salt.copyOf()
    val counterBytes = ByteBuffer.allocate(8).putLong(counter.toLong()).array()
    for (index in counterBytes.indices) {
        result[result.size - 8 + index] =
            (result[result.size - 8 + index].toInt() xor counterBytes[index].toInt()).toByte()
    }
    return result
}
